// mdb-cli/helpers/download-pro-starter.js
'use strict';

var config = require('../config');

var fs = require('fs');

var HttpWrapper = require('../utils/http-wrapper');

var unzip = require('unzipper');

var Path = require('path');

var ProgressBar = require('progress');

var _require = require('stream'),
    Readable = _require.Readable;

var CliStatus = require('../models/cli-status');

module.exports = {
  downloadProStarter: function downloadProStarter(packageName, headers, path, projectName) {
    return new Promise(function (resolve, reject) {
      var http = new HttpWrapper({
        port: config.port,
        hostname: config.host,
        path: "/packages/download/".concat(packageName),
        method: 'GET',
        data: '',
        headers: headers
      });
      var request = http.createRequest(function (response) {
        var result;
        var statusCode = response.statusCode,
            statusMessage = response.statusMessage;

        if (statusCode >= 400 && statusCode <= 500) {
          result = {
            'Status': statusCode,
            'Message': statusMessage
          };
          return reject(result);
        }

        var readStream = new Readable();

        readStream._read = function () {};

        var len = Number(response.headers['content-length']);
        var bar = new ProgressBar('[:bar] :eta s', {
          complete: '=',
          incomplete: ' ',
          width: 100,
          total: len
        });
        response.on('data', function (chunk) {
          readStream.push(chunk);
          bar.tick(chunk.length);
        });
        response.on('end', function () {
          result = [{
            'Status': CliStatus.SUCCESS,
            'Message': 'Initialization completed.'
          }];
          readStream.push(null);
          console.log('\n');
        });

        try {
          var tmpFolder = Path.join(path, 'tmp-mdb-projects-downloads-dir');
          readStream.pipe(unzip.Extract({
            path: tmpFolder
          })).on('close', function () {
            var toRename = Path.join(tmpFolder, packageName);
            var destination = Path.join(path, projectName);
            fs.rename(toRename, destination, function (err) {
              if (err) reject(err);else resolve(result);
            });
            fs.rmdirSync(tmpFolder, {
              recursive: true
            });
          });
        } catch (e) {
          result = {
            'Status': CliStatus.INTERNAL_SERVER_ERROR,
            'Message': 'Error initializing your project'
          };
          reject(result);
        }
      });
      request.on('error', reject);
      request.write('');
      request.end();
    });
  }
};
