// mdb-cli/utils/db-delete-handler.js
'use strict';

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var config = require('../config');

var inquirer = require('inquirer');

var helpers = require('../helpers');

var CliStatus = require('../models/cli-status');

var AuthHandler = require('./auth-handler');

var HttpWrapper = require('../utils/http-wrapper');

var DbDeleteHandler = /*#__PURE__*/function () {
  function DbDeleteHandler() {
    var authHandler = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new AuthHandler();

    _classCallCheck(this, DbDeleteHandler);

    this.authHandler = authHandler;
    this.database = null;
    this.databases = [];
    this.options = {
      port: config.port,
      hostname: config.host,
      path: '/databases',
      headers: this.authHandler.headers
    };
    this.result = [];
  }

  _createClass(DbDeleteHandler, [{
    key: "getResult",
    value: function getResult() {
      return this.result;
    }
  }, {
    key: "fetchDatabases",
    value: function () {
      var _fetchDatabases = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var http, databases;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                http = new HttpWrapper(this.options);
                _context.next = 3;
                return http.get();

              case 3:
                databases = _context.sent;
                this.databases = typeof databases === 'string' ? JSON.parse(databases) : databases;

                if (!(this.databases.length === 0)) {
                  _context.next = 7;
                  break;
                }

                return _context.abrupt("return", Promise.reject([{
                  Status: CliStatus.SUCCESS,
                  Message: 'You do not have any databases yet.'
                }]));

              case 7:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function fetchDatabases() {
        return _fetchDatabases.apply(this, arguments);
      }

      return fetchDatabases;
    }()
  }, {
    key: "askForDbName",
    value: function () {
      var _askForDbName = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
        var select;
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return inquirer.createPromptModule()([{
                  type: 'list',
                  name: 'name',
                  message: 'Choose database',
                  choices: this.databases
                }]);

              case 2:
                select = _context2.sent;
                this.database = this.databases.find(function (db) {
                  return db.name === select.name;
                });

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function askForDbName() {
        return _askForDbName.apply(this, arguments);
      }

      return askForDbName;
    }()
  }, {
    key: "confirmSelection",
    value: function () {
      var _confirmSelection = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3() {
        var name;
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                console.log('\n\x1b[31m%s\x1b[0m', ' Warning!', 'This operation cannot be undone. If it\'s a misclick you can abort the operation with Ctrl + C\n');
                _context3.next = 3;
                return helpers.showTextPrompt('Confirm deleting selected database by typing its name:', 'Database name must not be empty.');

              case 3:
                name = _context3.sent;

                if (!(name !== this.database.name)) {
                  _context3.next = 6;
                  break;
                }

                return _context3.abrupt("return", Promise.reject({
                  Status: CliStatus.CLI_ERROR,
                  Message: 'The names do not match.'
                }));

              case 6:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function confirmSelection() {
        return _confirmSelection.apply(this, arguments);
      }

      return confirmSelection;
    }()
  }, {
    key: "deleteDatabase",
    value: function () {
      var _deleteDatabase = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4() {
        var http;
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                this.options.path = "/databases/".concat(this.database.databaseId);
                http = new HttpWrapper(this.options);
                _context4.next = 4;
                return http["delete"]();

              case 4:
                this.result.push({
                  Status: CliStatus.HTTP_SUCCESS,
                  Message: 'Database successfully deleted.'
                });

              case 5:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function deleteDatabase() {
        return _deleteDatabase.apply(this, arguments);
      }

      return deleteDatabase;
    }()
  }]);

  return DbDeleteHandler;
}();

module.exports = DbDeleteHandler;
