// mdb-cli/utils/db-list-handler.js
'use strict';

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var config = require('../config');

var AuthHandler = require('./auth-handler');

var HttpWrapper = require('../utils/http-wrapper');

var DbListHandler = /*#__PURE__*/function () {
  function DbListHandler() {
    var authHandler = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new AuthHandler();

    _classCallCheck(this, DbListHandler);

    this.result = [];
    this.authHandler = authHandler;
    this.options = {
      port: config.port,
      hostname: config.host,
      path: '/databases',
      headers: this.authHandler.headers
    };
  }

  _createClass(DbListHandler, [{
    key: "getResult",
    value: function getResult() {
      return this.result;
    }
  }, {
    key: "fetchDatabases",
    value: function () {
      var _fetchDatabases = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var http, databases;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                http = new HttpWrapper(this.options);
                _context.next = 3;
                return http.get();

              case 3:
                databases = _context.sent;
                databases = typeof databases === 'string' ? JSON.parse(databases) : databases;

                if (!(databases.length === 0)) {
                  _context.next = 7;
                  break;
                }

                return _context.abrupt("return", this.result = [{
                  Status: 0,
                  Message: 'You do not have any databases yet.'
                }]);

              case 7:
                this.result = databases.map(function (db) {
                  return {
                    'Database': db.database,
                    'Name': db.name,
                    'Username': db.username,
                    'Connection String': db.connectionString,
                    'Description': db.description
                  };
                });

              case 8:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function fetchDatabases() {
        return _fetchDatabases.apply(this, arguments);
      }

      return fetchDatabases;
    }()
  }]);

  return DbListHandler;
}();

module.exports = DbListHandler;
