// mdb-cli/utils/login-handler.js
'use strict';

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var HttpWrapper = require('../utils/http-wrapper');

var AuthHandler = require('./auth-handler');

var CliStatus = require('../models/cli-status');

var helpers = require('../helpers/');

var config = require('../config');

var LoginHandler = /*#__PURE__*/function () {
  function LoginHandler(response) {
    var authHandler = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : new AuthHandler(false);

    _classCallCheck(this, LoginHandler);

    this.result = response;
    this.options = {
      port: config.port,
      hostname: config.host,
      path: '/auth/login',
      method: 'POST',
      data: '',
      headers: {
        'Content-Type': 'application/json'
      }
    };
    this._userToken = '';
    this.authHandler = authHandler;
  }

  _createClass(LoginHandler, [{
    key: "getResult",
    value: function getResult() {
      return this.result;
    }
  }, {
    key: "askCredentials",
    value: function askCredentials() {
      var _this = this;

      var prompt = require('inquirer').createPromptModule();

      return prompt([{
        type: 'text',
        message: 'Enter your MDB username',
        name: 'username',
        validate: function validate(value) {
          /* istanbul ignore next */
          var valid = Boolean(value);
          /* istanbul ignore next */

          return valid || 'Login must not be empty.';
        }
      }, {
        type: 'password',
        message: 'Enter your MDB password',
        name: 'password',
        mask: '*',
        validate: function validate(value) {
          /* istanbul ignore next */
          var valid = Boolean(value);
          /* istanbul ignore next */

          return valid || 'Password must not be empty.';
        }
      }]).then(function (answers) {
        var username = answers.username,
            password = answers.password;
        _this.options.data = JSON.stringify({
          username: username,
          password: password
        });
        _this.options.headers['Content-Length'] = Buffer.byteLength(_this.options.data);
      });
    }
  }, {
    key: "login",
    value: function login() {
      var http = new HttpWrapper(this.options);
      return http.post();
    }
  }, {
    key: "parseResponse",
    value: function parseResponse(response) {
      response = typeof response === 'string' ? JSON.parse(response) : response;

      var _response = response,
          _response2 = _slicedToArray(_response, 1),
          token = _response2[0].token;

      this._userToken = token;
    }
  }, {
    key: "saveToken",
    value: function saveToken() {
      try {
        var saved = helpers.saveToken(this._userToken);

        if (saved) {
          this.result = [{
            'Status': CliStatus.SUCCESS,
            'Message': 'Login successful'
          }];
        } else {
          this.result = [{
            'Status': CliStatus.ERROR,
            'Message': 'Login failed'
          }];
        }
      } catch (e) {
        this.result = [{
          'Status': CliStatus.INTERNAL_SERVER_ERROR,
          'Message': "Login failed: ".concat(e.message)
        }];
      }
    }
  }]);

  return LoginHandler;
}();

module.exports = LoginHandler;
