// mdb-cli/utils/unpublish-handler.js
'use strict';

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var config = require('../config');

var inquirer = require('inquirer');

var helpers = require('../helpers');

var AuthHandler = require('./auth-handler');

var CliStatus = require('../models/cli-status');

var HttpWrapper = require('../utils/http-wrapper');

var UnpublishHandler = /*#__PURE__*/function () {
  function UnpublishHandler() {
    var authHandler = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new AuthHandler();

    _classCallCheck(this, UnpublishHandler);

    this.result = [];
    this.projects = [];
    this.projectName = '';
    this.options = {
      port: config.port,
      hostname: config.host,
      path: '/project'
    };
    this.authHandler = authHandler;
    this.args = [];
    this.setAuthHeader();
  }

  _createClass(UnpublishHandler, [{
    key: "setAuthHeader",
    value: function setAuthHeader() {
      this.result = this.authHandler.result;
      this.options.headers = this.authHandler.headers;
    }
  }, {
    key: "setArgs",
    value: function setArgs(args) {
      this.args = args;
    }
  }, {
    key: "getResult",
    value: function getResult() {
      return this.result;
    }
  }, {
    key: "fetchProjects",
    value: function () {
      var _fetchProjects = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var http, projects;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                http = new HttpWrapper(this.options);
                _context.next = 3;
                return http.get();

              case 3:
                projects = _context.sent;
                projects = typeof projects === 'string' ? JSON.parse(projects) : projects;
                this.projects = projects.map(function (p) {
                  return {
                    name: p.projectName
                  };
                });

              case 6:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function fetchProjects() {
        return _fetchProjects.apply(this, arguments);
      }

      return fetchProjects;
    }()
  }, {
    key: "askForProjectName",
    value: function () {
      var _askForProjectName = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
        var select;
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (!(this.projects.length === 0)) {
                  _context2.next = 2;
                  break;
                }

                return _context2.abrupt("return", Promise.reject([{
                  Status: CliStatus.NOT_FOUND,
                  Message: 'You do not have any projects yet.'
                }]));

              case 2:
                if (!(this.args.length > 0)) {
                  _context2.next = 5;
                  break;
                }

                this.projectName = this.args[0];
                return _context2.abrupt("return");

              case 5:
                _context2.next = 7;
                return inquirer.createPromptModule()([{
                  type: 'list',
                  name: 'name',
                  message: 'Choose project',
                  choices: this.projects
                }]);

              case 7:
                select = _context2.sent;
                this.projectName = select.name;

              case 9:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function askForProjectName() {
        return _askForProjectName.apply(this, arguments);
      }

      return askForProjectName;
    }()
  }, {
    key: "confirmSelection",
    value: function () {
      var _confirmSelection = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3() {
        var name;
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return helpers.showTextPrompt('This operation cannot be undone. Confirm deleting selected project by typing its name:', 'Project name must not be empty.');

              case 2:
                name = _context3.sent;

                if (!(name !== this.projectName)) {
                  _context3.next = 5;
                  break;
                }

                return _context3.abrupt("return", Promise.reject({
                  Status: CliStatus.CLI_ERROR,
                  Message: 'The names do not match.'
                }));

              case 5:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function confirmSelection() {
        return _confirmSelection.apply(this, arguments);
      }

      return confirmSelection;
    }()
  }, {
    key: "unpublish",
    value: function () {
      var _unpublish = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4() {
        var http, result;
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                console.log("Unpublishing project ".concat(this.projectName, "..."));
                this.options.path = "/project/unpublish/".concat(this.projectName);
                http = new HttpWrapper(this.options);
                _context4.prev = 3;
                _context4.next = 6;
                return http["delete"]();

              case 6:
                result = _context4.sent;
                this.result = [{
                  'Status': CliStatus.HTTP_SUCCESS,
                  'Message': result
                }];
                _context4.next = 13;
                break;

              case 10:
                _context4.prev = 10;
                _context4.t0 = _context4["catch"](3);
                this.result = [{
                  'Status': _context4.t0.statusCode,
                  'Message': _context4.t0.message
                }];

              case 13:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this, [[3, 10]]);
      }));

      function unpublish() {
        return _unpublish.apply(this, arguments);
      }

      return unpublish;
    }()
  }]);

  return UnpublishHandler;
}();

module.exports = UnpublishHandler;
